src/base/preferences.h

```diff
@@ -188,6 +188,10 @@ class Preferences: public QObject
     void setWebUiHttpsCertificate(const QByteArray &data);
     QByteArray getWebUiHttpsKey() const;
     void setWebUiHttpsKey(const QByteArray &data);
+    bool isAltWebUiEnabled() const;
```

Comment:
These two new options should be merged. Probably with a "Set default" button in the UI.

Reply:
Initially I have only one option. But someone might want to turn "Alternative UI" off, but save the path for further use. Besides, I accept the idea that other settings can appear under "Alternative UI".

Reply:
The path can be saved elsewhere, what's the problem with that?

Reply:
Can't follow your thoughts...
We need one setting to enable/disable Alt UI, and the second one to store path.
How to disable it but store save path using only one setting?

Reply:
IMO, there is no need to save the path. Those who need it, can save it as they like. Or, even better, you may give them a list to choose from. Then there will be no need to save and re-type that path.

Reply:
>Or, even better, you may give them a list to choose from.

Where do I get this list, without saving anything?
>Then there will be no need to save and re-type that path.

It is not clear at all. I save it in order not to re-type it again next time user enable Alt UI.

Reply:
> Where do I get this list, without saving anything?

As far as I can see, the alternative UI path is relative. Contents of its parent directory is the list you need.

Reply:
>As far as I can see, the alternative UI path is relative.

You misunderstood this (what could bring you that idea?). It can be any path.

Reply:
> It can be any path.

Exceptionally bad decision (opens a wide way to security problems). Please change that.

Reply:
>Exceptionally bad decision (opens a wide way to security problems). Please change that.

Why???
User should decide himself where he want to put his files!
When you install e.g. Apache, does it forbid you to specify `www` folder you want?

Reply:
Wrong analogy. What you propose isto to change that path via HTTP request.

Reply:
>What you propose is to to change that path via HTTP request.

What do you mean?

Reply:
It is (or will be) possible to set alternative UI path via Web API. As such it is possible to load any file into qBt and compromise the machine it is running on.

Reply:
>It is (or will be) possible to set alternative UI path via Web API.

It isn't possible now and, of course, I wouldn't provide such a possibility anytime.
>it is possible to load any file into qBt and compromise the machine it is running on.

Any real scenario? I can't imagine it unless the machine is already compromised.

Reply:
And the next question. From your POV, what accepted path set can be here, that prevents possible security problems?

Reply:
In any case, I'm not a fan of "wipe the user ass". If something does not prevent the program to work logically, the user can decide how to use it.

Reply:
> Any real scenario?

Already here: https://github.com/qbittorrent/qBittorrent/pull/7610/files#diff-d00d2fcfdbe6ae0c443a24c3cccad6dbR205

> It isn't possible now and, of course, I wouldn't provide such a possibility anytime.

But how then users of the headless system should choose a web theme?

>  From your POV, what accepted path set can be here, that prevents possible security problems?

The most important here is that the paths set has to be limited and not include anything like /tmp and /var or any other directory, that can be populated during a normal system' workflow. It also may not include current directory (or app should set its working directory to the file system root dir). Also you can consider supplying this path via command line parameter (and then it may be an absolute path).

Reply:
> the paths set has to be limited and not include anything like /tmp and /var or any other directory, that can be populated during a normal system' workflow. It also may not include current directory (or app should set its working directory to the file system root dir).

IMO, it's up to the user, isn't it? Why it's "Wrong analogy" with the ordinary web-server app? I still can set it up to use any folder as `www` root (e.g. `/tmp` and so on, if I'm not right in the head or I have other good reasons to do it). Nobody forbids me to do it. That's my concern, if it will hurt the security of my machine.
Honestly, I don't want to mess with this. As soon as you start to do something like that, immediately there is someone who says: "Why do you tie my hands? It is very important for me (and for many others) to use that folder!".

Reply:
In any case, if you have more supporters, and you offer some reasonable way to limit this (or even better, implement it yourself), I won't fight it.
@sledgehammer999, @Chocobo1, @qbittorrent/webdev, @WolfganP?

Reply:
> "Why do you tie my hands? It is very important for me (and for many others) to use that folder!".

As I answered already, pass that directory via a command line parameter.

Reply:
Can you describe your point of view entirely, and not throw some broken up words?
And stop to answer only the part of question that is meaningless without the rest.

Reply:
My POV is:
The program is incorrect if it produces the incorrect output for correct input and settings. I'm willing (try) to correct the mistakes of this kind. But all you've presented so far, was as follows: "If the user set inappropriate folder as www root this will lead to errors". You propose to introduce any restrictions, but what? To try to "fix user brains" is "fight with windmills", IMO. Which folders should we ban? Only the user can know how exactly his filesystem is used. The only (more or less) reasonable restriction is to have only one predefined www folder (e.g. inside qBittorrent data dir).

Reply:
I'm supporting the freewill on webui root folder location, as I expect it will be used by somehow advanced users and the location may vary according to the setups (ie my qB is running in a headless raspi2, so most likely the webui files will end up in some ramdisk location after boot). 
Others may need a per user interface, and may choose /home/<somepath>...

Reply:
> The program is incorrect if it produces the incorrect output for correct input and settings.

I'm talking not about correctness of the program, because it is correctly implemented to contain the designed security hole.

Reply:
"Allow the user to do what he has a right" is a security hole?

Reply:
If a user is authenticated and passes a correct CSRF token with the request, what's the issue with allowing them to set the webui directory? At that point they have full control over the webui anyway, including the option to specify ANY external program to execute upon torrent completion.

Reply:
@Piccirello, shhh... If @evsh reads this, you'd be anathematized and burned at the stake!

Reply:
Yes, any binary can be downloaded to the target system and executed via post-download command. Yet I strongly believe that presence of a mistake or design flaw in a program may not be used as an excuse for adding more problems.